### ISSUE TITLE
N64 Controller Control Stick range not working properly with Nintendo Switch Pro controller or adapters

### ISSUE THREAD

**Issue by Mousecopp** [OpenEmu/OpenEmu#3102]
I've been using Mupen64Plus on OpenEmu on Mac for months now, and even after updating to 2.0.5, Mario doesn't move faster than a few units per second.  Until now, I've had little to no issues, but now all of a sudden the control stick inputs barely move the character at all in any game I try.

**Comment by clobber** [comment]
Which gamepad? You are saying all of a sudden, as in when? Are you saying it used to work in a certain version and now it doesn't? When did the issue first occur?

Are you using OpenEmu 2.0.5 with the Mupen64Plus core plugin version 2.5.3?

**Comment by Tundra77** [comment]
I'm having a similar problem. Playing Ocarina of Time on Mupen64Plus and was working fine. Suddenly after the update today, it doesn't seem to be registering the control stick range (i.e. Link no longer runs, just walks everywhere). I'm using a Nintendo Switch Pro Controller. Note: it was working fine pre-update.

**Comment by clobber** [comment]
Cannot replicate this with my DualShock 4 or SteelSeries Nimbus.

@Tundra77 What happens if you delete your bindings file? Please try the following:
1. Close the app
2. Navigate to the folder `~/Library/Application Support/OpenEmu/Bindings` (~ meaning your Home folder where ~ is your username. Guide here if you don't know this: http://osxdaily.com/2016/12/12/show-user-library-folder-macos-sierra/)
3. Delete `Default.oebindings` in the "Bindings" folder.
4. Restart the app.

**Comment by Tundra77** [comment]
Tried that. No change. One other thing I noticed - not sure if relevant, but it seems to work when pushing to the left on the left analog stick, just not in any other direction. 100% sure it's not a controller issue, was literally working up to the second before I accepted the install of the update and not since...

**Comment by Tundra77** [comment]
Also just tried Majora's mask and having the exact same issue there.

**Comment by clobber** [comment]
Okay, thanks, but it's not going to be game-specific. It's an issue with analog range values passed from the Switch Pro controller to the frontend.

**Comment by Mousecopp** [comment]
Yeah, apologies on my wording and comment, it wasn't very specific.  When I said suddenly, I meant that I booted up the emulator in the same fashion I always do, and it just wasn't working.  I first had the issue when using Mupen64Plus on OpenEmu 2.0.4, and even after updating to OE 2.0.5, with Mupen64Plus 2.5.3, the issue persisted.  I'm still using the same Mayflash 2 port USB converter to play on a normal nintendo-made n64 controller.  I've tried different controllers and ROMs as well and the same problem occurs.  I also tried deleting Default.oebindings, but nothing changed.  Even though the mapping was empty, the control stick range still isn't registering.  Do you think it could just be a problem that's occurring with the Mayflash?

**Comment by Mousecopp** [comment]
And I'm assuming I wouldn't want to get rid of the entire Bindings folder, correct?

**Comment by Cau64** [comment]
currently having the same problem. ever since the recent update, was working perfectly fine before. some help would be great.

**Comment by clobber** [comment]
@Cau64 Actually, you can help by posting your gamepad instead of a vague "me too" reply.

**Comment by Cau64** [comment]
Apologies, I Am using a gameseek 2-port n64 adaptor with a genuine n64 controller on mac.

**Comment by salsolatragus** [comment]
I'm having the same issue. Using an original N64 controller with a [Mayflash N64 Controller Adapter](https://www.amazon.com/MAYFLASH-N64-Controller-Adapter-Pc/dp/B002B9FIUU). Tried two different controllers, Zelda - Ocarina of Time, Zelda - Majora's Mask, and downgrading to 2.0.4 (replacing the app and deleting ~/Library/Application Support/OpenEmu), but still Link is creeping along... it's the same whether I keep or delete my savegames. Is there something else I should consider for a downgrade?

**Comment by clobber** [comment]
Please don't "downgrade" the application - you will end up with more problems and doing that has nothing to do with solving this particular edge case.

The problem is some adapters and devices are not reporting the full analog range, normalized from 0.0 to 1.0, so when the Mupen64Plus core receives the value and multiplies axisValue * 80 (0 - 80 is the range of N64 analog stick), it is not enough to move the stick where it ought to really be.

The fact that is "worked" in a previous release of the Mupen64Plus core plugin is merely on accident, and didn't really "work" - it would break analog in some games such as Golden Eye and others on gamepads that properly report the full range (see the extensive list of gamepads I tested above). So it was an incorrect implementation.

There is not yet a fix in for these gamepads and adapters that don't appear to be reporting the analog values correctly. You can view the offending commit here: https://github.com/OpenEmu/Mupen64Plus-Core/commit/44dd073396e104c32d8dd5768b427ec9c93fe775

**Comment by salsolatragus** [comment]
Since the fix is in the Mupen core, is it possible to use the old core fore the time being?

**Comment by Mousecopp** [comment]
@clobber Thanks for your help, dude.  May sound sarcastic, but it actually isn't lol.

**Comment by TheSuperJez** [comment]
@salsolatragus  can you share the old core please?

**Comment by salsolatragus** [comment]
Here you go https://www.dropbox.com/sh/m9pk52i93zhyryj/AAChT3IRFjqAqCAY-tmMn2eba?dl=0

**Comment by clobber** [comment]
Please don't link to core plugins from third parties which could be altered. Link directly to the Mupen64Plus-Core repo Releases if you must.

**Comment by salsolatragus** [comment]
Sorry, I didn't know there is such a repo. I guess in that case there's no need for a link since everybody can figure it out. Thanks for the hint ;)

**Comment by dponskiy** [comment]
I am currently am having the same exact issue, mario/link wont run in their games just walk.. Any ideas on how to fix it??

**Comment by kylegordy** [comment]
I'm also having the same joystick range issue mentioned above. I'm using the [Mayflash USB Adaptor](https://www.amazon.com/MAYFLASH-N64-Controller-Adapter-Pc/dp/B002B9FIUU) along with an OEM N64 controller. Let me know if any more details about my setup might help.

**Comment by ghostinajar** [comment]
I am experiencing this issue with Nintendo Switch Pro Controller and OpenEmu 2.0.5

**Comment by usualsuspects2** [comment]
@salsolatragus's solution fixed this issue for me (with the Mayflash USB Adaptor). Just in case anyone else finds this thread who also isn't super adept at this sort of thing:
Here's the link: https://github.com/OpenEmu/Mupen64Plus-Core/releases
Download Mupen64Plus 2.5.2, then just replace the Mupen64Plus.oecoreplugin that's currently in the Core folder with this older version.

**Comment by HandTheUnicorn** [comment]
Sorry for bumping an old, closed dohicky, but for anyone else googling how to get a Nintendo Switch Pro controller to work properly with OpenEmu, @salsoatragus and @usualsuspects2 right above me pretty much figured it out. In case you're as stupid as I am, though, the Mupen64Plus.oecoreplugin goes in ~/System/Application Support/OpenEmu/Cores and not /Applications/OpenEmu/Plugins/Systems. It's a major pain that you have to actually turn off your internet every time instead of opting out of the new versions of everything, but it'll get your joysticks to work.

**Comment by TheSuperJez** [comment]
@HandTheUnicorn unfortunately, i think nobody is gonna fix this, as it is not problem with the emulator itself, as stick ranges in Nintendo controllers are lower,  so the recommended "workaround" is playing Goldeneye 64, with a "proper" PS4 controller instead any other from Nintendo (good look).
Even with retroarch that has more development activity, sticks won't work correctly on OSX, the only way i have found in order to play with these nintendo sticks, is playing in Windows with project 64, after calibrating the joystick in control panel, they work like charm (at least with N64 controller)

**Comment by clobber** [comment]
@TheSuperJez aka "Master of Science" and Programmer - learn to contribute something (maybe some code?) instead of noise.

**Comment by clobber** [comment]
Hi all,

I am working to solve this in the short term for the **Wii U Pro** and **Switch Pro** gamepads. Unfortunately, I'm not focusing on the N64 adapters right now - there are far too many which makes it hard to hack around the issue. The general problem is these gamepads are not made for PC use and are not giving you access to the full analog range. For example, a device will report `0` as the minimum and `65535` as the maximum values, but as you push the stick all the way it never reaches close to `65535` because it is physically blocked by the plastic circle around the stick. The long term fix will be a manual calibration method which will also take some UI/UX work.

So for those with Nintendo Wii U Pro and Switch Pro gamepads, please do the following:

1) Make sure you have no software running or installed that can interfere with input as outlined in the [input troubleshooting guide](https://github.com/OpenEmu/OpenEmu/wiki/Troubleshooting:-Input-problems).

2) Download the test build https://transfer.sh/HUOQu/OpenEmu-inputtest5.zip

3) Pair/connect your controller.

4) Open the test build.

5) Open Console.app, go to the search filter box at the top and type openemu, then click Clear.

6) Go back to OpenEmu and move your **left** analog stick in a full circle several times, making sure you fully touch the edges of the side of the plastic ring as far as it can go. Then do the same for the **right** analog stick.

7) Press "Button A" on your controller.

8) Go back to Console.app and copy all of that output and paste it back here. You will see:
```
------begin axis log------
blah
blah
blah
blah
blah
------end axis log------
```

**Comment by spacecowgoesmoo** [comment]
I downloaded the test build and I think I followed the instructions, but I'm not getting any axis log output in Console.app.

**Comment by clobber** [comment]
@spacecowgoesmoo Which gamepad, and is it genuine or a clone/knockoff?

**Comment by spacecowgoesmoo** [comment]
It's a genuine Wii U Pro controller. My OS X version is 10.11.6.

**Comment by clobber** [comment]
It is my understanding that "Button A" on the Wii U Pro corresponds to HID Button 33 from its profile. I don't know why it wouldn't be outputting if that is the case. Try pressing all other buttons on the gamepad to see if it logs.

**Comment by clobber** [comment]
Furthermore, verify that input actually works in-game with that build and your device paired.

**Comment by spacecowgoesmoo** [comment]
My inputs are working fine in-game. I tested all the buttons including the analog stick buttons, but still no console messages.

**Comment by shysaur** [comment]
@clobber The patched build you posted does not work as intended, because you patched ``-OE_setupEventWithDeviceHandler:value:`` but the events dispatched by ``OEWiimoteHIDDeviceHandler`` are not initialized through that method. (Tested with a Wii U pro controller)

The safest place to patch seems to be ``-OE_dispatchAxisEventWithAxis:minimum:value:maximum:timestamp:cookie:`` and the similar function for buttons in ``OEWiimoteHIDDeviceHandler``.

I can try making a patched build myself, if you want, as I already reverse-engineered your patch enough to know what it did, and I can also test it.

**Comment by clobber** [comment]
@shysaur Thanks, indeed I did patch that method. I forgot the Wii U Pro also goes through that special device handler. You can probably patch `buttonEventWithDeviceHandler` and `axisEventWithDeviceHandler` in OEHIDEvent though.

**Comment by shysaur** [comment]
True, but that would also be called for other kinds of controller other than the ones being tested (and this can contaminate the data). In this case I don't think it's very likely, but better safe than sorry.

**Comment by shysaur** [comment]
Try this build:
https://transfer.sh/vEEBU/OpenEmu_inputtest_wii.zip
The procedure is the same as before; except that you don't have to open Console.app. A dialog box will show up with the information to paste here when you press A (you can select the text in the dialog box and use command-C and command-V to copy and paste it).

It would be also be interesting to also see some centering information; my controller for example has a small offset at rest. To get that, do the same procedure as before, but don't touch the analog sticks before pushing the A button.

Every time the A button is pressed, the old measurement data is reset; in this way it's easier to try again from scratch if needed.

@clobber I added a "samples collected" counter; it gets incremented every time an event is sent.

---

Data from my controller
Full range:
```
------begin axis log------
manstr: Unknown prodstr: Nintendo RVL-CNT-01-UC vid: 1406 pid: 816
left  analog X axis min: -1394 max: 866
left  analog Y axis min: -1335 max: 1013
right analog X axis min: -1320 max: 962
right analog Y axis min: -1311 max: 1005
# samples collected: 28460
------end axis log------
```
At rest:
```
------begin axis log------
manstr: Unknown prodstr: Nintendo RVL-CNT-01-UC vid: 1406 pid: 816
left  analog X axis min: -225 max: -220
left  analog Y axis min: -147 max: -142
right analog X axis min: -154 max: -150
right analog Y axis min: -134 max: -130
# samples collected: 3164
------end axis log------
```

**Comment by clobber** [comment]
Here's the updated build I was just about to post: https://transfer.sh/pXNPw/OpenEmu_inputtest6.zip

@shysaur Just an FYI, your build is packaged with experimental/WIP system plugins that should really be scrubbed out before uploading, unless we want to give people false hope :P

**Comment by shysaur** [comment]
> I was just about to post

I took the "You can probably patch..." sentence as an implicit "go ahead"; sorry for that.

> unless we want to give people false hope :P

They could have clicked on the OpenEmu project page on github to get that false hope every day of the year for several years now, and more easily than snooping on bug reports that they don't care about.

> I should note that inputtest6 build also logs the 'logical' min/max ranges that get reported, but those might look strange for the Wii U Pro. I guess we will see!

**Comment by shysaur** [comment]
@clobber mmmh I'm getting only (null) values from your build :-(
(I'll also be more careful when I select the scheme the next time I make a build to be posted here)
Specimen:
```
2017-11-17 23:10:49.359 OpenEmu[25935:1218341] ------begin axis log------
2017-11-17 23:10:49.359 OpenEmu[25935:1218341] manstr: Unknown prodstr: Nintendo RVL-CNT-01-UC vid: 1406 pid: 816
2017-11-17 23:10:49.359 OpenEmu[25935:1218341] left  analog X axis min: (null) max: (null) (logical min: -1250 max 1250)
2017-11-17 23:10:49.359 OpenEmu[25935:1218341] left  analog Y axis min: (null) max: (null) (logical min: -1250 max 1250)
2017-11-17 23:10:49.359 OpenEmu[25935:1218341] right analog X axis min: (null) max: (null) (logical min: -1250 max 1250)
2017-11-17 23:10:49.359 OpenEmu[25935:1218341] right analog Y axis min: (null) max: (null) (logical min: -1250 max 1250)
2017-11-17 23:10:49.360 OpenEmu[25935:1218341] ------end axis log------
```

**Comment by clobber** [comment]
>They could have clicked on the OpenEmu project page on github to get that false hope every day of the year for several years now, and more easily than snooping on bug reports that they don't care about.

I'm not totally sure how to take that, but I went ahead and removed them from your build and reuploaded it/edited it in your post anyway. Please don't take that the wrong way. I prefer if we are not distributing builds with those system plugins because I like to manage expectations since we note in one of the wiki pages what is alpha/WIP and reasons for why they may never be available. I just don't want anyone getting the wrong impression.

>I'm getting only (null) values from your build 

Hrm that's weird. I was able to get output from my Wii with Wii Pro attached. Let's use your build for testing the Wii U Pro, but can you add the logging of the min/max logical values too as I had? It's important we see those so we know just how much of a difference there is. If you post an updated build I will move it to the original comment https://github.com/OpenEmu/OpenEmu/issues/3102#issuecomment-345170009 and mark it as "Wii U Pro testing only"

**Comment by shysaur** [comment]
> I'm not totally sure how to take that, but I went ahead and removed them from your build and reuploaded it/edited it in your post anyway.

I was simply trying to find an excuse to feel less bad about myself. Thank you for reuploading a sanitized build yourself! My upload speed is rather slow and I appreciate not having to wait 20 minutes again just for that.

> can you add the logging of the min/max logical values too as I had? 
sure

**Comment by clobber** [comment]
Whoops I see what was wrong there. Attached is a fixed build for the null logging: https://transfer.sh/mrJRK/OpenEmu_inputtest7.zip (Note: In Console it will spam a lot of invalid output when you first pair your Wii U Pro).

**Comment by shysaur** [comment]
By the way, on second thought, why do you say we need to log the min/max logical ranges? If I'm not misinterpreting what you are saying, the min/max logical ranges for the pro controller are constants equal to -1250 and +1250 respectively, and for the wii remote (nunchuck) they are -128 and +127 respectively. It can be easily found by looking at the code.

Your last build works! Results:
```
2017-11-18 00:02:12.590 OpenEmu[28551:1523961] ------begin axis log------
2017-11-18 00:02:12.590 OpenEmu[28551:1523961] manstr: Unknown prodstr: Nintendo RVL-CNT-01-UC vid: 1406 pid: 816
2017-11-18 00:02:12.594 OpenEmu[28551:1523961] left  analog X axis min: -1404 max: 866 (logical min: -1250 max 1250)
2017-11-18 00:02:12.598 OpenEmu[28551:1523961] left  analog Y axis min: -1335 max: 1010 (logical min: -1250 max 1250)
2017-11-18 00:02:12.603 OpenEmu[28551:1523961] right analog X axis min: -1315 max: 962 (logical min: -1250 max 1250)
2017-11-18 00:02:12.607 OpenEmu[28551:1523961] right analog Y axis min: -1327 max: 1009 (logical min: -1250 max 1250)
2017-11-18 00:02:12.607 OpenEmu[28551:1523961] ------end axis log------
```

**Comment by clobber** [comment]
>why do you say we need to log the min/max logical ranges

So the thing is those values currently should only pertain to the Wiimote stuff, because the range on the Wii U Pro appears to be different (we are effectively capping the ranges at -1250 to 1250). With your device, you are hitting observed ranges of -1400 to 1010. So because the highest you can hit is ~860-1000 max, Mario can never reach top speed running because we say the actual max is 1250 (it's not!). Mario will just end up walking all the time.

Since the Wii U Pro shares the device handler class and reuses a lot of code as the Wiimote+attachments, I think this problem can be fixed a bit easier than the Switch controller. We just have to set a more appropriate min/max range (and possibly deadzone value) for the Wii U Pro so that when `OE_dispatchAxisEventWithAxis` is called, it sends the proper `maximum`, `minimum` and `value` off to OEHIDEvent.

I don't have this gamepad to experiment with, but with a bit more data (hopefully from @spacecowgoesmoo) we should know the baseline ranges to use for the Wii U Pro.

**Comment by spacecowgoesmoo** [comment]
Results, from inputtest7.zip. I did about 10 rotations per analog stick.

`11/17/17 3:37:55.398 PM OpenEmu[15597]: ------begin axis log------
11/17/17 3:37:55.398 PM OpenEmu[15597]: manstr: Unknown prodstr: Nintendo RVL-CNT-01-UC vid: 1406 pid: 816
11/17/17 3:37:55.406 PM OpenEmu[15597]: left  analog X axis min: -1161 max: 1094 (logical min: -1250 max 1250)
11/17/17 3:37:55.413 PM OpenEmu[15597]: left  analog Y axis min: -1313 max: 988 (logical min: -1250 max 1250)
11/17/17 3:37:55.421 PM OpenEmu[15597]: right analog X axis min: -1279 max: 1013 (logical min: -1250 max 1250)
11/17/17 3:37:55.428 PM OpenEmu[15597]: right analog Y axis min: -1277 max: 1050 (logical min: -1250 max 1250)
11/17/17 3:37:55.428 PM OpenEmu[15597]: ------end axis log------`

**Comment by clobber** [comment]
@shysaur I don't mean that Wiimote code is being used incorrectly, nor do I have any reason to not believe you. I'm just saying the ranges being used are not good enough for actual observed ranges on a Wii U Pro. +/-1250 isn't cutting it. We both understand this so what is there to argue over? Anyway, since I don't have this device to test with I leave it up to others to decide more accurate defaults per-axes. It does look like it never gets too far beyond 1000 max or approaches right up to it, so that is probably a good place to start to make Mario run again :)

**Comment by shysaur** [comment]
@clobber I didn't want to bring about any drama; but from your response it seemed to me that you were saying something else. My deepest apologies.

@spacecowgoesmoo Great Job!

Could you please download the other build (https://transfer.sh/UJJoU/OpenEmu_inputtest_wii.zip), open it, connect your controller, wait for a few seconds without touching it, and pressing A? Then paste here the contents of the dialog box that appears. I'm really curious about the values that your controller returns when the sticks are in the center position. (The version you have already downloaded would not work as it rounds these values to zero.)

Both our controllers have a reduced range on the X axis of the left stick, and now I wonder if it's true for other controllers. Anyway, at this point I would say that a candidate range that stays on the safe side is -1000/+800.

**Comment by spacecowgoesmoo** [comment]
Controller #1
```
------begin axis log------
manstr: Unknown prodstr: Nintendo RVL-CNT-01-UC vid: 1406 pid: 816
left  analog X axis min: 38 max: 43
left  analog Y axis min: -152 max: -147
right analog X axis min: -142 max: -137
right analog Y axis min: -70 max: -66
# samples collected: 4896
------end axis log------
```

Controller #2:
```
------begin axis log------
manstr: Unknown prodstr: Nintendo RVL-CNT-01-UC vid: 1406 pid: 816
left  analog X axis min: -172 max: -170
left  analog Y axis min: -167 max: -156
right analog X axis min: -85 max: -82
right analog Y axis min: -153 max: -147
# samples collected: 3656
------end axis log------
```

**Comment by shysaur** [comment]
@spacecowgoesmoo Thank you a lot! One last thing: could you please do the rotation test also with the second controller you have? The more data we collect, the better it is!

If I find the time tomorrow I will post a build with modified controller parameters, so that we can do some real life "testing" 😉 🎮.

**Comment by spacecowgoesmoo** [comment]
@shysaur I already did, I added it to my other post a few minutes after testing the first controller.

**Comment by shysaur** [comment]
@spacecowgoesmoo You are right! For some reason I couldn’t see it before you replied. I was trusting github’s auto page refresh a bit too much it seems.

**Comment by shysaur** [comment]
Ok; I have a test build with improved ranges for the **Wii U Pro Controller**. You can download it from here: https://transfer.sh/OAS0D/OpenEmu_wiiupro_test.zip (the link will expire in 14 days).

Give it a try and please reply to this thread with feedback (whether it's positive or negative). Make sure to test how the controller behaves when the analog sticks are used as digital buttons (i.e., for the N64 C-pad).

@clobber The code changes I did are here: shysaur/OpenEmu-SDK@2bebd68. If no issues are discovered, I'll make a PR.

**Comment by spacecowgoesmoo** [comment]
Using OpenEmu_wiiupro_test.zip:

Controller 1:
Ratchet & Clank PSP: Analog movement range is greatly improved, but still has a tiny bit of deadzone.
Dissidia 012: Great improvement, no deadzone at all.
Untold Legends 2, FF7 Crisis Core: Good improvement, but it wasn't too bad to begin with. Tiny bit of deadzone.
Mario 64: Good improvement, no deadzone.
The left analog (in this build, and not in the current release) has a slight neutral drift to the right most of the time. Easy to test in Mario 64's opening menu.
N64 c-stick behavior seems a bit more finicky than usual. Slightly easier to trigger C-UP when I think I'm pressing C-LEFT or C-RIGHT

Controller 2:
Ratchet & Clank PSP, Dissidia 012: Great improvement, no deadzone.
Untold Legends 2, FF7 Crisis Core, Mario 64: Good improvement, no deadzone.
No analog stick drift on this controller. It does sit in my closet most of the time..
N64 c-stick behavior still finicky.

**Comment by shysaur** [comment]
It turns out that I did a mistake when coding the changes and it probably caused the tiny deadzones the drift and the C-stick finicky-ness. An updated build will be up shortly.

**Comment by spacecowgoesmoo** [comment]
I should have clarified that the deadzones I mentioned don't apply to the entire stick. The available ranges vary by direction, so on Controller 1 I can move at full speed if my Y axis is neutral or pointing up, and then I'll slow down as my analog stick gets closer to a certain point in the lower left.

**Comment by shysaur** [comment]
Updated build: https://transfer.sh/27iW2/OpenEmu_wiiupro_test2.zip

I looked a bit more closely at the C-stick situation. With the new build it should be a bit less finicky, but it's probably still a bit too easy to hit two directions simultaneously by accident. But -- just from looking at the code -- it seems to me that it's a controller-independent problem.

The root issue seems that there is no analog threshold for activating digital buttons bound to analog controls (in other words, if a digital control is bound to an analog control, the digital control will be unpressed when the analog value is exactly zero, and pressed in all other situations). But I am sure that @clobber knows a lot more about this than me, and maybe there is a reason for this behaviour.

**Comment by shysaur** [comment]
> I should have clarified that the deadzones I mentioned don't apply uniformly to the entire stick.

In the last build this *should* not happen anymore, unless I screwed up again.

This is in general a behaviour that seems to be in common to every Wii U Pro Controller we have data from, most probably because the inner PCB that supports the analog sticks is not perfectly centred in respect to the hole for the sticks in the outer shield.

**Comment by spacecowgoesmoo** [comment]
Using OpenEmu_wiiupro_test2.zip:

Controllers 1 & 2:
- Dissidia 012, FF7 Crisis Core, Mario 64, Ratchet & Clank PSP, Untold Legends 2: Analog range is perfect, no deadzone <3
- No analog drift.
- N64 c-stick is good. Still slightly weird by nature, but it seems to be back to normal.

**Comment by shysaur** [comment]
Great! At this point I'm pretty confident that the new parameters are correct, and they are also very similar to the ones used by the Windows tool [WiinUPro](https://github.com/KeyPuncher/WiinUPro/blob/master/Nintroller/Calibrations.cs). I'm gonna create a PR right now (note that it won't be visible in the OpenEmu project, but in the OpenEmu-SDK project).

This issue will probably remain open, as @clobber also wanted to investigate the Switch Pro controller, and no progress has been done on that so far.

**Comment by Ghoom** [comment]
With OpenEmu_inputtest7.zip on Switch Pro controller

------begin axis log------
manstr: Unknown prodstr: Pro Controller vid: 1406 pid: 8201
left  analog X axis min: 7248 max: 55408 (logical min: 0 max 65535)
left  analog Y axis min: 11695 max: 62047 (logical min: 0 max 65535)
right analog X axis min: 6992 max: 55168 (logical min: 0 max 65535)
right analog Y axis min: 10287 max: 59839 (logical min: 0 max 65535)
------end axis log------

**Comment by clobber** [comment]
@Ghoom Thanks. So your y-axes are weird which I've seen before. Give the following steps a try, which I've been told can help 'reset' the range and then run the test again to see if your numbers are any different:

1) Go to the Bluetooth preferences, right click the Pro Controller and select "Disconnect"
2) Plug the controller in via USB
3) Unplug the controller
4) Reconnect the Pro Controller via Bluetooth
5) If the y-axes numbers don't look any better, also try completely removing the the device in Bluetooth preferences and then pair it again.

**Comment by Ghoom** [comment]
Done.

------begin axis log------
manstr: Unknown prodstr: Pro Controller vid: 1406 pid: 8201
left  analog X axis min: 7344 max: 55312 (logical min: 0 max 65535)
left  analog Y axis min: 11695 max: 61935 (logical min: 0 max 65535)
right analog X axis min: 6992 max: 55104 (logical min: 0 max 65535)
right analog Y axis min: 10303 max: 59839 (logical min: 0 max 65535)
------end axis log------

**Comment by Ghoom** [comment]
No calibrate on ysjoystickreader.

<img width="793" alt="ysjoystickreader" src="https://user-images.githubusercontent.com/29138038/33041926-4de93db6-ce40-11e7-9ecc-9be7cd855bcc.png">

**Comment by clobber** [comment]
Your odd y-axes ranges did not improve. I think I will have to eventually buy the Switch Pro Controller to better investigate this.

**Comment by clobber** [comment]
@Tundra77, @ghostinajar, @HandTheUnicorn and other Switch Pro Controller owners that discover this issue, please follow https://github.com/OpenEmu/OpenEmu/issues/3102#issuecomment-345170009 because we need more data.

**Comment by HandTheUnicorn** [comment]
Oh jeeze, I forgot about this. I read through (fascinating stuff! You guys are doing a great thing for people, here) and saw nobody posted about the Nintendo _SWITCH_ Pro Controller, so here's my dump of that:

```
------begin axis log------
manstr: Unknown prodstr: Pro Controller vid: 1406 pid: 8201
left  analog X axis min: 6832 max: 54912 (logical min: 0 max 65535)
left  analog Y axis min: 8111 max: 58287 (logical min: 0 max 65535)
right analog X axis min: 6208 max: 54432 (logical min: 0 max 65535)
right analog Y axis min: 8479 max: 59087 (logical min: 0 max 65535)
------end axis log------
```



...looking again, there's another Switchy user right above me.

**Comment by juanfm288** [comment]
Hi, here is my dump: 

------begin axis log------
manstr: Unknown prodstr: Pro Controller vid: 1406 pid: 8201
left  analog X axis min: 7152 max: 54992 (logical min: 0 max 65535)
left  analog Y axis min: 9247 max: 58799 (logical min: 0 max 65535)
right analog X axis min: 6272 max: 53760 (logical min: 0 max 65535)
right analog Y axis min: 11215 max: 58863 (logical min: 0 max 65535)
------end axis log------

Thanks for such amazing app.

**Comment by davidtessier** [comment]
Thank you in advance for looking into this. You're doing fantastic work!

```
------begin axis log------
manstr: Unknown prodstr: Pro Controller vid: 1406 pid: 8201
left  analog X axis min: 4768 max: 55232 (logical min: 0 max 65535)
left  analog Y axis min: 10079 max: 61887 (logical min: 0 max 65535)
right analog X axis min: 29184 max: 29280 (logical min: 0 max 65535)
right analog Y axis min: 33935 max: 34015 (logical min: 0 max 65535)
------end axis log------
```

**Comment by clobber** [comment]
@davidtessier As per the directions, you are supposed to do both left stick and right stick in the same test as everyone else has. Not separately.

**Comment by zachcoyle** [comment]
```
------begin axis log------
manstr: Unknown prodstr: Pro Controller vid: 1406 pid: 8201
left  analog X axis min: 10944 max: 60352 (logical min: 0 max 65535)
left  analog Y axis min: 7135 max: 57775 (logical min: 0 max 65535)
right analog X axis min: 8192 max: 57808 (logical min: 0 max 65535)
right analog Y axis min: 9727 max: 60079 (logical min: 0 max 65535)
------end axis log------
```

**Comment by masterhand789** [comment]
Here are three tests from my genuine Nintendo Switch Pro Controller. There's a bit of variance between them since the measurement isn't perfect. If you want more data ask. Right now I'm trying to use this information to configure the controller in the Mupen64Plus emulator directly, outside of OpenEmu.

```
------begin axis log------
manstr: Unknown prodstr: Pro Controller vid: 1406 pid: 8201
left  analog X axis min: 8352 max: 57600 (logical min: 0 max 65535)
left  analog Y axis min: 10351 max: 60815 (logical min: 0 max 65535)
right analog X axis min: 9536 max: 58208 (logical min: 0 max 65535)
right analog Y axis min: 7375 max: 58911 (logical min: 0 max 65535)
------end axis log------
```

```
------begin axis log------
manstr: Unknown prodstr: Pro Controller vid: 1406 pid: 8201
left  analog X axis min: 8336 max: 57456 (logical min: 0 max 65535)
left  analog Y axis min: 10239 max: 60751 (logical min: 0 max 65535)
right analog X axis min: 9472 max: 58096 (logical min: 0 max 65535)
right analog Y axis min: 7407 max: 58927 (logical min: 0 max 65535)
------end axis log------
```

```
------begin axis log------
manstr: Unknown prodstr: Pro Controller vid: 1406 pid: 8201
left  analog X axis min: 8320 max: 57520 (logical min: 0 max 65535)
left  analog Y axis min: 10383 max: 60735 (logical min: 0 max 65535)
right analog X axis min: 9456 max: 58048 (logical min: 0 max 65535)
right analog Y axis min: 7359 max: 58927 (logical min: 0 max 65535)
------end axis log------
```

**Comment by redvelociraptor** [comment]
I've also run into this problem. My results from two tests on my Nintendo Switch Pro Controller using the input test version posted on 12/20. 
```
------begin axis log------
manstr: Unknown prodstr: Pro Controller vid: 1406 pid: 8201
left  analog X axis min: 5152 max: 54448 (logical min: 0 max 65535)
left  analog Y axis min: 8191 max: 59999 (logical min: 0 max 65535)
right analog X axis min: 5600 max: 54784 (logical min: 0 max 65535)
right analog Y axis min: 8671 max: 60719 (logical min: 0 max 65535)
------end axis log------
```

```
------begin axis log------
manstr: Unknown prodstr: Pro Controller vid: 1406 pid: 8201
left  analog X axis min: 4944 max: 54464 (logical min: 0 max 65535)
left  analog Y axis min: 8127 max: 60031 (logical min: 0 max 65535)
right analog X axis min: 5616 max: 55344 (logical min: 0 max 65535)
right analog Y axis min: 8591 max: 60815 (logical min: 0 max 65535)
------end axis log------
```

**Comment by Marszene** [comment]
@clobber 
> Contributing guidelines - "Please search before opening a new issue."
#3102

I'm sorry but I've read this thread before I wrote mine... I did not find any solution (if there's one)...it seems like you are still working on the problem collecting data from other Switch users.

So do you confirm that today there's still no solution and that we have to wait?

**Comment by Endreho** [comment]
Has there been any sort of progress with fixing this? Thus far, i've been using Controllermate as a adjustment, but with that one, i'm only able to sprint

**Comment by clobber** [comment]
@Endreho provide useful diagnostic information instead of noise. https://github.com/OpenEmu/OpenEmu/issues/3102#issuecomment-345170009

**Comment by Endreho** [comment]
With a Nintendo Switch Pro Controller

```
------begin axis log------
manstr: Unknown prodstr: Pro Controller vid: 1406 pid: 8201
left  analog X axis min: 6704 max: 55632 (logical min: 0 max 65535)
left  analog Y axis min: 9535 max: 59999 (logical min: 0 max 65535)
right analog X axis min: 6656 max: 55024 (logical min: 0 max 65535)
right analog Y axis min: 10831 max: 61135 (logical min: 0 max 65535)
------end axis log------
```

**Comment by HandTheUnicorn** [comment]
@clobber Sorry, I can't find a private message option and I've been trying everything for a good couple of hours, now. The post of mine you deleted on Dec 19 had instructions on how to downgrade to a version where the Switch Pro Controller worked correctly, and had no other issues _for me_ in the games I want to play. I've switched computers and can't find old Mupen64 cores for OpenEmu anywhere, and I don't even know what versions I'd need anyway. Can you E-mail a copy of the deleted comment to me? I know you're making great progress with all the updates that there have been, but I just want to play two N64 games.

**Comment by clobber** [comment]
Reuploaded and updated the link. $80 is $60 too much for this gamepad btw, where Switch owners are complaining about [issues with the d-pad](https://imgur.com/gallery/Jruwv).

**Comment by mayfrogs** [comment]
Here are two tests using a Switch pro controller. Thanks for all your hard work

```
------begin axis log------
manstr: Unknown prodstr: Pro Controller vid: 1406 pid: 8201
left  analog X axis min: 7984 max: 55488 (logical min: 0 max 65535)
left  analog Y axis min: 8687 max: 59647 (logical min: 0 max 65535)
right analog X axis min: 31792 max: 31872 (logical min: 0 max 65535)
right analog Y axis min: 34719 max: 34799 (logical min: 0 max 65535)
------end axis log------
```
```
------begin axis log------
manstr: Unknown prodstr: Pro Controller vid: 1406 pid: 8201
left  analog X axis min: 7888 max: 55344 (logical min: 0 max 65535)
left  analog Y axis min: 8671 max: 59647 (logical min: 0 max 65535)
right analog X axis min: 31792 max: 31872 (logical min: 0 max 65535)
right analog Y axis min: 34719 max: 34799 (logical min: 0 max 65535)
------end axis log------
```

**Comment by Nathanael8** [comment]
Nintendo switch pro controller
```
------begin axis log------
manstr: Unknown prodstr: Pro Controller vid: 1406 pid: 8201
left  analog X axis min: 7520 max: 56592 (logical min: 0 max 65535)
left  analog Y axis min: 8639 max: 60927 (logical min: 0 max 65535)
right analog X axis min: 4672 max: 54128 (logical min: 0 max 65535)
right analog Y axis min: 9039 max: 60959 (logical min: 0 max 65535)
------end axis log------
```

**Comment by oitsjustjose** [comment]
So I hate to be one to nag, but is there a fix for this? I have confirmed the same issue on RetroArch using Parallel and Mupen64, and on OpenEmu using Mupen64. This is an issue on the new A-Power Gamecube Bluetooth controllers, so it's honestly quite irritating to see this be an issue across multiple great controllers. Seems like the issue is just that the states for these controllers are unlike many.

**Comment by clobber** [comment]
@oitsjustjose 

>So I hate to be one to nag

Let's not carelessly reply to an open issue, please, where dozens of people will now receive a needless email.

> but is there a fix for this?

If it were fixed this issue would be closed with a commit referenced.

>Seems like the issue is

Toward the beginning of the thread describes the problem, you can go and read it.

**Comment by clobber** [comment]
Fixed in https://github.com/OpenEmu/OpenEmu-SDK/compare/d3b2cc9ab818...ce01c3fd1576

For those tracking this issue, please test the following build making sure it resolves your issue with the Nintendo Switch Pro Controller analog sticks and N64 games. Leave feedback in this thread.

**Note: To test, only connect your Nintendo Switch Pro Controller via Bluetooth, not USB.**

Download: https://transfer.sh/HUZeO/OpenEmu-switch_pro_fix.zip

**Comment by yougotwill** [comment]
@clobber Tested with Pro Controller works perfectly! Had to redo the control mappings in preferences but after that everything worked perfectly. Thanks for the hard work!

**Comment by clobber** [comment]
@yougotwill Thanks for testing. To clarify:

The problem is resolved in the test build versus the stock 2.0.9.1 release? Just want to confirm you can actually replicate an analog problem in N64 games with stock 2.0.9.1 release. Also, you said you had to redo your control mappings - any reason for that? We automap controls for this device, so I'm curious.

**Comment by TerryAg** [comment]
Works perfectly for me (except the website to download it, lots of "your connection is not private" and Chrome not letting me go to the page). I didn't have to redo any control mappings.

**Comment by clobber** [comment]
Thanks for testing.

@TerryAg Looks like a bug with the new render after the resolution changes, specifically for a PAL region game. That will be fixed in time for next release.

**Comment by stuartcarnie** [comment]
@TerryAg I just pushed fixes to resolve the rendering issue

**Comment by jflatby** [comment]
@clobber The link you provided with a fix for this is dead. 
And the problem still remains on the official downloadable version, as you know. (2.0.9.1)

Is there a way to implement the fix from that commit without downloading 17 git repositories and building with xcode etc.? Or do I need to go through all that?

Thanks.

**Comment by bopob** [comment]
Link died again. Please fix.

**Comment by clobber** [comment]
Thanks for the interest. We have already confirmed this as fixed from multiple testers now, so I don’t think we require more. The fix will be available in the next OpenEmu release.
